### PR TITLE
[SPARK-4413][SQL] Parquet support through datasource API

### PR DESCRIPTION
**Pull request by marmbrus**
Goals:
- Support for accessing parquet using SQL but not requiring Hive (thus allowing support of parquet tables with decimal columns)
- Support for folder based partitioning with automatic discovery of available partitions
- Caching of file metadata

See scaladoc of `ParquetRelation2` for more details.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 1]
```diff
@@ -93,3 +98,18 @@ abstract class PrunedScan extends BaseRelation {
...
+ * An interface for experimenting with a more direct connection to the query planner.  Compared to
+ * [[PrunedFilteredScan]], this operator receives the raw expressions from the
+ * [[org.apache.spark.sql.catalyst.plans.logical.LogicalPlan]].  Unlike the other APIs this
+ * interface is not designed to be binary compatible across releases and thus should only be used
+ * for experimentation.
+ */
+@Experimental
+abstract class CatalystScan extends BaseRelation {
+  def buildScan(requiredColumns: Seq[Attribute], filters: Seq[Expression]): RDD[Row]
+}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Do we really need this? Currently none of our build-in data source use this and there is no example to demonstrate how to use it.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: I believe it's OK to remove this one now. I thought it had been removed.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Why remove it?  Its labeled experimental, its not broken...

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Oh, yeah, now I remembered that we had once discussed about this and decided to keep it there for future experimentation. Sorry that I totally forgot that...

**Pull request closed without merging**